tests/api/compiler.cpp

```diff
@@ -212,15 +212,160 @@ TEST_F(WithCommandQueue, ModuleScopeConstantData) {
+    EnqueueReadBuffer(buffer, CL_BLOCKING, 0, buffer_size, result);
+
+    ASSERT_EQ(result[0], 0);
+    ASSERT_EQ(result[1], 1);
+    ASSERT_EQ(result[2], 2);
+    ASSERT_EQ(result[3], 3);
+}
+
+TEST_F(WithCommandQueue, ProgramBinaryLinkLibrary) {
+    static const char* source = R"(
+      kernel void test(global uint *output) {
+        uint gid = get_global_id(0);
+        output[gid] = gid;
+      }
+    )";
+    auto program = CreateProgram(source);
+    CompileProgram(program);
+    const size_t gws = 4;
+    const size_t buffer_size = gws * sizeof(cl_uint);
+    auto buffer = CreateBuffer(CL_MEM_WRITE_ONLY, buffer_size);
+    cl_uint result[gws] = {0};
+    ASSERT_EQ(buffer_size, sizeof(result));
+
+    cl_program program_list = program;
+    auto linked_program = LinkProgram(1, &program_list, "-create-library");
+    ASSERT_EQ(GetProgramBinaryType(linked_program),
+              CL_PROGRAM_BINARY_TYPE_LIBRARY);
+    auto linked_binary = GetProgramBinary(linked_program);
+    auto binary_program = CreateProgramWithBinary(linked_binary);
+    ASSERT_EQ(GetProgramBinaryType(binary_program),
+              CL_PROGRAM_BINARY_TYPE_LIBRARY);
+
+    BuildProgram(binary_program);
+    auto kernel = CreateKernel(binary_program, "test");
+    SetKernelArg(kernel, 0, buffer);
+
+    EnqueueNDRangeKernel(kernel, 1, nullptr, &gws, nullptr);
+    EnqueueReadBuffer(buffer, CL_BLOCKING, 0, buffer_size, result);
+
+    ASSERT_EQ(result[0], 0);
+    ASSERT_EQ(result[1], 1);
+    ASSERT_EQ(result[2], 2);
+    ASSERT_EQ(result[3], 3);
+}
+
+TEST_F(WithCommandQueue, ProgramBinaryLink) {
+    static const char* source = R"(
+      kernel void test(global uint *output) {
+        uint gid = get_global_id(0);
+        output[gid] = gid;
+      }
+    )";
+    auto program = CreateProgram(source);
+    CompileProgram(program);
+    const size_t gws = 4;
+    const size_t buffer_size = gws * sizeof(cl_uint);
+    auto buffer = CreateBuffer(CL_MEM_WRITE_ONLY, buffer_size);
+    cl_uint result[gws] = {0};
+    ASSERT_EQ(buffer_size, sizeof(result));
+
+    cl_program program_list = program;
+    auto linked_program = LinkProgram(1, &program_list);
+    ASSERT_EQ(GetProgramBinaryType(linked_program),
+              CL_PROGRAM_BINARY_TYPE_EXECUTABLE);
+    auto linked_binary = GetProgramBinary(linked_program);
+    auto binary_program = CreateProgramWithBinary(linked_binary);
+    ASSERT_EQ(GetProgramBinaryType(binary_program),
+              CL_PROGRAM_BINARY_TYPE_EXECUTABLE);
+
+    BuildProgram(binary_program);
+    auto kernel = CreateKernel(binary_program, "test");
+    SetKernelArg(kernel, 0, buffer);
+
+    EnqueueNDRangeKernel(kernel, 1, nullptr, &gws, nullptr);
+    EnqueueReadBuffer(buffer, CL_BLOCKING, 0, buffer_size, result);
+
+    ASSERT_EQ(result[0], 0);
+    ASSERT_EQ(result[1], 1);
+    ASSERT_EQ(result[2], 2);
+    ASSERT_EQ(result[3], 3);
+}
+
+TEST_F(WithCommandQueue, ProgramBinaryExecutable) {
```

Comment:
All these tests have a lot in common. It might be worth introducing a dedicated fixture.